cwl/src/main/scala/cwl/CommandOutputExpression.scala

```diff
@@ -0,0 +1,133 @@
+package cwl
+
+import common.validation.ErrorOr.ErrorOr
+import wom.expression.IoFunctionSet
+import wom.types._
+import wom.values.{WomArray, WomFile, WomGlobFile, WomMap, WomString, WomValue}
+import cats.syntax.validated._
+import cats.syntax.either._
+
+import common.validation.Validation._
+import cats.data.NonEmptyList
+
+import scala.language.postfixOps
+import scala.concurrent.Await
+import scala.concurrent.duration._
+import mouse.all._
+
+case class CommandOutputExpression(outputBinding: CommandOutputBinding,
+                                   override val cwlExpressionType: WomType,
+                                   override val inputs: Set[String]) extends CwlWomExpression {
+
+  // TODO WOM: outputBinding.toString is probably not the best representation of the expression source
+  override def sourceString = outputBinding.toString
+
+  override def evaluateValue(inputValues: Map[String, WomValue], ioFunctionSet: IoFunctionSet): ErrorOr[WomValue] = {
+
+    val parameterContext = ParameterContext().addInputs(inputValues)
+
+    /*
+    CommandOutputBinding.glob:
+    Find files relative to the output directory, using POSIX glob(3) pathname matching. If an array is provided, find
+    files that match any pattern in the array. If an expression is provided, the expression must return a string or an
+    array of strings, which will then be evaluated as one or more glob patterns. Must only match and return files which
+    actually exist.
+
+    http://www.commonwl.org/v1.0/CommandLineTool.html#CommandOutputBinding
+     */
+    def outputBindingEvaluationResult: Either[NonEmptyList[String], WomValue] = {
+      import StringOrExpression._
+      (outputBinding, parameterContext) match {
+        case (CommandOutputBinding(_, _, Some(String(value))), Right(_)) => WomString(value).asRight
+        case (CommandOutputBinding(Some(glob), _, None), Right(parameterContext)) =>
+          GlobEvaluator.globPaths(glob, parameterContext, ioFunctionSet) match {
+            case Seq(fileName) => WomString(fileName).asRight
+            case array => WomArray(WomArrayType(WomStringType), array.map(WomString.apply)).asRight
+          }
+
+
+        case (CommandOutputBinding(glob, loadContents, Some(Expression(expression))), Right(parameterContext)) =>
+
+          val paths: Seq[String] = glob.toSeq flatMap { globValue =>
+            GlobEvaluator.globPaths(globValue, parameterContext, ioFunctionSet)
+          }
+
+          val _loadContents: Boolean = loadContents getOrElse false
+
+          val womMaps: Array[Map[String, String]] =
+            paths.toArray map {
+              (path: String) =>
+                // TODO: WOM: basename/dirname/size/checksum/etc.
+                val contents =
+                  Map("contents" -> load64KiB(path, ioFunctionSet)).
+                    filter(_ => _loadContents)
```

Comment:
This is still reading even if we don't want the contents and I think is more complicated that it needs to be.